pageserver/src/tenant/mgr.rs

```diff
@@ -279,24 +280,44 @@ pub async fn create_tenant(
     ctx: &RequestContext,
 ) -> Result<Arc<Tenant>, TenantMapInsertError> {
     tenant_map_insert(tenant_id, |vacant_entry| {
+        let succeed = Cell::new(false);
+
         // We're holding the tenants lock in write mode while doing local IO.
         // If this section ever becomes contentious, introduce a new `TenantState::Creating`
         // and do the work in that state.
         let tenant_directory = super::create_tenant_files(conf, tenant_conf, tenant_id, CreateTenantFilesMode::Create)?;
-        // TODO: tenant directory remains on disk if we bail out from here on.
-        //       See https://github.com/neondatabase/neon/issues/4233
+
+        scopeguard::defer! {
+            if !succeed.get() {
+                std::fs::remove_dir_all(&tenant_directory).ok(); // ignore the error
```

Comment:
It would feel more natural to log the rare case of this remove_dir_all failing with `warn!` for example.